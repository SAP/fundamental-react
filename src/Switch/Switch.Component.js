import path from 'path';
import React from 'react';
import { Switch } from '..';
import { ComponentPage, Example } from '../_playground';

export const SwitchComponent = () => {
    return (
        <ComponentPage
            description={`The **Switch** component is used to activate or deactivate an element. It uses a visual metaphor that is known
                to the user with visible differences between on and off state. It is recommended to always display the
                Switch with a label above it as well as the label of the selected state. For example, the label above
                would be "Active", the Switch state would be “on” and the selected state label displayed to the right of
                the Switch would be “Yes”.`}
            sourceModulePath={path.join(__dirname, './Switch')}
            title='Switch'>

            <Example title='Sizes'>
                <>
                    <Switch>
                        Default Switch
                    </Switch>
                    <Switch compact>
                        Compact Switch
                    </Switch>
                </>
            </Example>

            <Example title='Disabled State'>
                <>
                    <Switch disabled>
                        Disabled Switch
                    </Switch>
                    <Switch checked disabled>
                        Disabled Checked Switch
                    </Switch>
                </>
            </Example>

            <Example title='Semantic'>
                <Switch
                    internalLabels={{
                        checked: {
                            text: 'on',
                            glyph: 'accept'
                        },
                        unchecked: {
                            text: 'off',
                            glyph: 'decline'
                        }
                    }}
                    semantic>Semantic Switch with Icons</Switch>
            </Example>
            <Example title='Internal labels'>
                <>
                    <Switch
                        internalLabels={{
                            checked: {
                                text: 'on'
                            },
                            unchecked: {
                                text: 'off'
                            }
                        }}
                        showInternalLabels>Show Internal Labels</Switch>
                    <Switch
                        internalLabels={{
                            checked: {
                                text: 'on',
                                glyph: 'accept'
                            },
                            unchecked: {
                                text: 'off',
                                glyph: 'decline'
                            }
                        }}
                        semantic
                        showInternalLabels>Show Internal Labels and Icons</Switch>
                </>
            </Example>


        </ComponentPage>
    );
};
