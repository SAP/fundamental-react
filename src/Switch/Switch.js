import classnames from 'classnames';
import CustomPropTypes from '../utils/CustomPropTypes/CustomPropTypes';
import FormLabel from '../Forms/FormLabel';
import keycode from 'keycode';
import PropTypes from 'prop-types';
import SwitchItem from './_SwitchItem';
import React, { useCallback, useEffect, useState } from 'react';


const Switch = React.forwardRef(({
    checked,
    children,
    compact,
    disabled,
    disableStyles,
    className,
    id,
    inputProps,
    internalLabels,
    localizedText,
    onChange,
    semantic,
    showInternalLabels,
    ...props
}, ref) => {

    useEffect(() => {
        if (!disableStyles) {
            require('fundamental-styles/dist/icon.css');
            require('fundamental-styles/dist/switch.css');
        }
    }, []);

    let [isChecked, setIsChecked] = useState(!!checked);

    const handleChange = (e) => {
        setIsChecked(!isChecked);
        onChange(e);
    };

    const onKeyDownSwitch = useCallback((event) => {
        switch (keycode(event)) {
            case 'enter':
                event.preventDefault();
                event.stopPropagation();
                setIsChecked(!isChecked);
                break;
            default:
        }
    });

    const spanClasses = classnames(
        'fd-switch',
        {
            'fd-switch--compact': compact,
            'fd-switch--semantic': semantic
        }
    );

    let internalLabelDisplay;
    if (internalLabels && showInternalLabels) {
        const internalLabelText = isChecked ? internalLabels.checked.text : internalLabels.unchecked.text;
        internalLabelDisplay = <span aria-live='polite' className='fd-switch__text'>{internalLabelText}</span>;
    }

    return (
        <>
            <FormLabel
                disabled={disabled}
                {...props}
                disableStyles={disableStyles}
                htmlFor={id}
                onKeyDown={onKeyDownSwitch}>
                {children}
            </FormLabel>
            <label className='fd-switch__label'>
                {internalLabelDisplay}
                <span className={spanClasses}>
                    <input
                        {...inputProps}
                        aria-checked={isChecked}
                        aria-label={localizedText.switchLabel}
                        checked={isChecked}
                        className='fd-switch__input'
                        disabled={disabled}
                        id={id}
                        onChange={handleChange}
                        ref={ref}
                        type='checkbox' />
                    <div className='fd-switch__wrapper'>
                        <div className='fd-switch__track'>
                            {internalLabels ? (
                                <>
                                    <SwitchItem
                                        glyph={internalLabels.checked.glyph}
                                        text={internalLabels.checked.text}
                                        type='on' />
                                    <span className='fd-switch__handle' role='presentation' />
                                    <SwitchItem
                                        glyph={internalLabels.unchecked.glyph}
                                        text={internalLabels.unchecked.text}
                                        type='off' />
                                </>
                            ) : (<span className='fd-switch__handle' role='presentation' />)
                            }
                        </div>
                    </div>
                </span>
            </label>
        </>
    );

});


Switch.displayName = 'Switch';

Switch.propTypes = {
    checked: PropTypes.bool,
    children: PropTypes.node,
    className: PropTypes.string,
    compact: PropTypes.bool,
    disabled: PropTypes.bool,
    disableStyles: PropTypes.bool,
    id: PropTypes.string,
    inputProps: PropTypes.object,
    internalLabels: PropTypes.shape({
        checked: PropTypes.shape(SwitchItem.PropTypes),
        unchecked: PropTypes.shape(SwitchItem.PropTypes)
    }),
    labelProps: PropTypes.object,
    localizedText: CustomPropTypes.i18n({
        switchLabel: PropTypes.string
    }),
    semantic: PropTypes.bool,
    showInternalLabels: PropTypes.bool,
    onChange: PropTypes.func
};

Switch.defaultProps = {
    localizedText: {
        switchLabel: 'Switch'
    },
    onChange: () => { }
};

Switch.propDescriptions = {
    checked: 'Set to true for component to be checked on render.',
    internalLabels: 'Provide text and/or an icon for labels inside the Switch component.',
    showInternalLabels: 'Set to true to display text from `internalLabels` next to the switch.'
};

export default Switch;
