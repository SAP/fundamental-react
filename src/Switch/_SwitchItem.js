import classnamesBind from 'classnames/bind';
import Icon from '../Icon/Icon';
import { listOfIcons } from '../utils/listOfIcons';
import PropTypes from 'prop-types';
import React from 'react';
import withStyles from '../utils/withStyles';
import iconStyles from 'fundamental-styles/dist/icon.css';
import switchStyles from 'fundamental-styles/dist/switch.css';

const classnames = classnamesBind.bind({
    ...iconStyles,
    ...switchStyles
});

const SwitchItem = ({ glyph, text, type, cssNamespace }) => {

    const iconClasses = classnames(
        {
            [`${cssNamespace}-switch__icon`]: !!glyph,
            [`${cssNamespace}-switch__icon--${type}`]: !!glyph
        }
    );

    return (
        <Icon
            ariaLabel={text}
            className={iconClasses}
            glyph={glyph} />
    );

};

SwitchItem.displayName = 'SwitchItem';

SwitchItem.propTypes = {
    /** Localized text for the label */
    text: PropTypes.string.isRequired,
    /** Optional icon to include. See the icon page for the list of icons */
    glyph: PropTypes.oneOf(listOfIcons),
    /** Sets the variation of the component. Primarily used for styling */
    type: PropTypes.oneOf(['on', 'off'])
};

export default withStyles(SwitchItem);
