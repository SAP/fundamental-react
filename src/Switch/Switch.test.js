import { mount } from 'enzyme';
import React from 'react';
import Switch from './Switch';

describe('<Switch />', () => {
    const defaultSwitch = <Switch>Normal Switch</Switch>;


    test('Switch state change', () => {
        const wrapper = mount(defaultSwitch);
        // check that Switch is not checked
        expect(wrapper.find('input').props().checked).toBe(false);
        expect(wrapper.find('input').props()['aria-checked']).toBe(false);

        // simulate changing Switch
        wrapper.find('input[type="checkbox"]').simulate('change');

        // check that Switch is checked
        expect(wrapper.find('input').props().checked).toBe(true);
        expect(wrapper.find('input').props()['aria-checked']).toBe(true);
    });

    describe('Switch default rendering', () => {
        test('should default to a not checked state', () => {
            const element = mount(<Switch />);

            expect(element.find('input').props().checked).toBe(false);
            expect(element.find('input').props()['aria-checked']).toBe(false);
        });
        test('should have truthy checked state when passed checked prop', () => {
            const element = mount(<Switch checked />);

            expect(element.find('input').props().checked).toBe(true);
            expect(element.find('input').props()['aria-checked']).toBe(true);
        });
    });

    describe('onChange handler', () => {
        test('should dispatch the onChange callback with the event', () => {
            let f = jest.fn();
            const element = mount(<Switch data-sample='Sample' onChange={f} />);

            element.find('input[type="checkbox"]').simulate('change');

            expect(f).toHaveBeenCalledTimes(1);
        });
    });

    describe('Prop spreading', () => {
        test('should allow props to be spread to the Switch component', () => {
            const element = mount(<Switch data-sample='Sample' />);

            expect(
                element.find('label').at(0).getDOMNode().attributes['data-sample'].value
            ).toBe('Sample');
        });

        test('should allow props to be spread to the Switch component\'s input element', () => {
            const element = mount(<Switch inputProps={{ 'data-sample': 'Sample' }} />);

            expect(
                element.find('input').getDOMNode().attributes['data-sample'].value
            ).toBe('Sample');
        });
    });

    describe('Switch state change on props update', () => {
        const wrapper = mount(<Switch checked />);

        test('should change to not checked state if new props are sent with checked as false', () => {
            //updating prop to checked as false
            wrapper.setProps({ checked: false });
            wrapper.update();

            // check that switch is not checked
            expect(wrapper.find('input').props().checked).toBe(false);
            expect(wrapper.find('input').props()['aria-checked']).toBe(false);
        });

        test('should change to checked state if new props are sent with checked as true', () => {
            //updating prop to checked as true
            wrapper.setProps({ checked: true });
            wrapper.update();

            // check that switch is checked
            expect(wrapper.find('input').props().checked).toBe(true);
            expect(wrapper.find('input').props()['aria-checked']).toBe(true);
        });
    });
});
