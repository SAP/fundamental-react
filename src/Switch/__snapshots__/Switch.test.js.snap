// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Switch /> create Switch component 1`] = `
Array [
  <label
    className="fd-form-label"
    onKeyDown={[Function]}
  >
    Normal Switch
  </label>,
  <label
    className="fd-switch__label"
  >
    <span
      className="fd-switch"
    >
      <input
        aria-checked={false}
        aria-label="Switch"
        checked={false}
        className="fd-switch__input"
        onChange={[Function]}
        type="checkbox"
      />
      <div
        className="fd-switch__wrapper"
      >
        <div
          className="fd-switch__track"
        >
          <span
            className="fd-switch__handle"
            role="presentation"
          />
        </div>
      </div>
    </span>
  </label>,
]
`;

exports[`<Switch /> create Switch component 2`] = `
Array [
  <label
    className="fd-form-label"
    onKeyDown={[Function]}
  >
    Compact Switch
  </label>,
  <label
    className="fd-switch__label"
  >
    <span
      className="fd-switch fd-switch--compact"
    >
      <input
        aria-checked={true}
        aria-label="Switch"
        checked={true}
        className="fd-switch__input"
        onChange={[Function]}
        type="checkbox"
      />
      <div
        className="fd-switch__wrapper"
      >
        <div
          className="fd-switch__track"
        >
          <span
            className="fd-switch__handle"
            role="presentation"
          />
        </div>
      </div>
    </span>
  </label>,
]
`;

exports[`<Switch /> create Switch component 3`] = `
Array [
  <label
    className="fd-form-label is-disabled"
    onKeyDown={[Function]}
  >
    Normal Switch
  </label>,
  <label
    className="fd-switch__label"
  >
    <span
      className="fd-switch"
    >
      <input
        aria-checked={false}
        aria-label="Switch"
        checked={false}
        className="fd-switch__input"
        disabled={true}
        onChange={[Function]}
        type="checkbox"
      />
      <div
        className="fd-switch__wrapper"
      >
        <div
          className="fd-switch__track"
        >
          <span
            className="fd-switch__handle"
            role="presentation"
          />
        </div>
      </div>
    </span>
  </label>,
]
`;

exports[`<Switch /> create Switch component 4`] = `
Array [
  <label
    className="fd-form-label is-disabled"
    onKeyDown={[Function]}
  >
    Normal Switch
  </label>,
  <label
    className="fd-switch__label"
  >
    <span
      className="fd-switch"
    >
      <input
        aria-checked={true}
        aria-label="Switch"
        checked={true}
        className="fd-switch__input"
        disabled={true}
        onChange={[Function]}
        type="checkbox"
      />
      <div
        className="fd-switch__wrapper"
      >
        <div
          className="fd-switch__track"
        >
          <span
            className="fd-switch__handle"
            role="presentation"
          />
        </div>
      </div>
    </span>
  </label>,
]
`;

exports[`<Switch /> create Switch component 5`] = `
Array [
  <label
    className="fd-form-label"
    onKeyDown={[Function]}
  >
    Show Internal Labels
  </label>,
  <label
    className="fd-switch__label"
  >
    <span
      aria-live="polite"
      className="fd-switch__text"
    >
      on
    </span>
    <span
      className="fd-switch"
    >
      <input
        aria-checked={true}
        aria-label="Switch"
        checked={true}
        className="fd-switch__input"
        onChange={[Function]}
        type="checkbox"
      />
      <div
        className="fd-switch__wrapper"
      >
        <div
          className="fd-switch__track"
        >
          <span
            aria-label="on"
            className="fd-fd-switch__text"
          />
          <span
            className="fd-switch__handle"
            role="presentation"
          />
          <span
            aria-label="off"
            className="fd-fd-switch__text"
          />
        </div>
      </div>
    </span>
  </label>,
]
`;

exports[`<Switch /> create Switch component 6`] = `
Array [
  <label
    className="fd-form-label"
    onKeyDown={[Function]}
  >
    Show Internal Labels
  </label>,
  <label
    className="fd-switch__label"
  >
    <span
      aria-live="polite"
      className="fd-switch__text"
    >
      off
    </span>
    <span
      className="fd-switch"
    >
      <input
        aria-checked={false}
        aria-label="Switch"
        checked={false}
        className="fd-switch__input"
        onChange={[Function]}
        type="checkbox"
      />
      <div
        className="fd-switch__wrapper"
      >
        <div
          className="fd-switch__track"
        >
          <span
            aria-label="on"
            className="fd-fd-switch__text"
          />
          <span
            className="fd-switch__handle"
            role="presentation"
          />
          <span
            aria-label="off"
            className="fd-fd-switch__text"
          />
        </div>
      </div>
    </span>
  </label>,
]
`;
